engagementlab/engagement-lab-website-2.0
# Docker Containerization

## Roadmap

- [x] Create Dockerfile for `/client`
- [x] Create Dockerfile for `/server`
- [x] Create `docker-compose.yml` to orchestrate client and server containers
- [ ] ~~Push images to Docker Hub~~
- [ ] ~~Use `docker-compose` environment variables instead of `dotenv` `.env` file~~

**Note:** Pushing images to Docker Hub is not needed since images can be built via `docker-compose`
**Note:** `docker-compose` can read the `.env` file from the server the same way `dotenv` would, so setting all the environment variables in `docker-compose.yml` isn't needed.

## Updates in this branch

- Uses latest (unpubished) version of [el-bootstrapper](https://github.com/engagementlab/el-bootstrapper) with `MONGO_ADDRESS` environment variable
- Add `preconnect` and `dns-prefetch` directives to html files for faster page loading
- Change `href` schema from `https://<url>` to `//<url>` since https requests to the API fail during docker development
  - This can be mitigated in production using [nginx 301 rules](https://serverfault.com/a/337893)


## Testing

0. Set up Nginx (see below)
1. Copy `server/.env.example` to `server/.env` and fill out fields
2. In the root directory of the repo, run `docker-compose up --build`
3. Once the containers are spun up, visit http://elhome.local:8080, the site should load normally.


## Nginx

Nginx is needed to redirect `/api` requests to the api container.  This has only been tested with Docker for Mac + the standard homebrew Nginx installation.

```nginx
upstream client {
	server 127.0.0.1:3001;
}

upstream api {
	server 127.0.0.1:3000;
}

server {
	listen 8080;
	# Add this to /etc/hosts so it points to 127.0.0.1
	server_name elhome.local;

	location / {
		proxy_read_timeout 10s;
		proxy_pass http://client;
	}

	location ^~ /api/ {
		proxy_read_timeout 10s;
		proxy_pass http://api;
	}
}

```
